blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubEnterpriseScm.java

```diff
@@ -30,14 +35,46 @@ public GithubEnterpriseScm(Reachable parent) {
     }
 
     @Override
+    public @Nonnull String getUri() {
+        String apiUri = getCustomApiUri();
+
+        // NOTE: GithubEnterpriseScm requires that the apiUri be specified
+        if (StringUtils.isEmpty(apiUri)) {
+            throw new ServiceException.BadRequestException(new ErrorMessage(400, "URI is required"));
+        }
+
+        return apiUri;
+    }
+
+    @Override
+    public String getCredentialId() {
+        // there is no "default" credential for GitHub Enterprise, so return null
+        return null;
+    }
+
+    @Override
     public String getCredentialDomainName() {
-        java.net.URI uri;
-        try {
-            uri = new URI(getUri());
-        } catch (URISyntaxException e) {
-            throw new ServiceException.UnexpectedErrorException(new ErrorMessage(400, "Invalid Github Enterprise URI: "+getUri()));
+        return DOMAIN_NAME;
+    }
+
+    @Exported(name = CREDENTIALS)
+    public List<GithubEnterpriseScmCredential> getCredentials() {
```

Comment:
@cliffmeyers Hmm, I think its not going to fit in here. The thing is if we know apiUrl then it should return at most single element in the list, that is credential for the given apiUrl.  Then that means, `getCredentialId()` would return that same credential anyways. Basically we do not need 'credentials' element in the response in that case.

I think we missed important aspect in our discussion, that is if apiUrl is optional in case of GHE, then Scm abstraction doesn't work as it is, because it returns single scm url. See `Scm.getUri()`.  In that case we need collection of Scm in case of GHE, each on to represent instance of GHE.

I think we are better of calling Credential Api which can give list of credentials for given user (user store) and domain. If there is any issue in that API we can fix such query capability but I am fairly certain its all there. 

My take on this after more thinking is that we should not implement `GithubEnterpriseScm.getCredentials()`.